jira/jira.go

```diff
@@ -1,41 +1,101 @@
 package jira
 
 import (
+	gojira "github.com/andygrunwald/go-jira"
 	"github.com/go-chat-bot/bot"
 	"os"
 	"regexp"
 	"strings"
 )
 
 const (
-	pattern = "(^|\\s)+[A-z]{3}-[0-9]+\\b"
-	env     = "JIRA_ISSUES_URL"
+	pattern    = ".*?([A-Z]+)-([0-9]+)\\b"
+	env        = "JIRA_ISSUES_URL"
+	userEnv    = "JIRA_USER"
+	passEnv    = "JIRA_PASS"
+	baseURLEnv = "JIRA_BASE_URL"
 )
 
 var (
-	url string
-	re  = regexp.MustCompile(pattern)
+	url      string
+	baseURL  string
+	jiraUser string
+	jiraPass string
+	projects map[string]gojira.Project
+	client   *gojira.Client
+	re       = regexp.MustCompile(pattern)
 )
 
-func getIssue(text string) string {
-	if re.MatchString(text) {
-		return re.FindString(text)
+func getProjects() map[string]gojira.Project {
+	req, _ := client.NewRequest("GET", "rest/api/2/project", nil)
+
+	projectObjects := new([]gojira.Project)
+	_, err := client.Do(req, projectObjects)
+	if err != nil {
+		panic(err)
+	}
+	projects = make(map[string]gojira.Project)
+	for _, project := range *projectObjects {
+		projects[project.Key] = project
+	}
+	return projects
+}
+
+func getIssues(text string) [][2]string {
+	matches := re.FindAllStringSubmatch(text, -1)
+	var data [][2]string
+	for _, match := range matches {
+		// match[1] == project key
+		// match[2] == issue number
+		data = append(data, [2]string{match[1], match[2]})
 	}
-	return ""
+	return data
 }
 
-func jira(cmd *bot.PassiveCmd) (string, error) {
-	issue := getIssue(cmd.Raw)
-	if issue != "" {
-		return url + strings.ToUpper(strings.TrimSpace(issue)), nil
+func jira(cmd *bot.PassiveCmd) (bot.CmdResultV3, error) {
+	result := bot.CmdResultV3{
+		Message: make(chan string),
+		Done:    make(chan bool, 1)}
+	result.Channel = cmd.Channel
+	issues := getIssues(cmd.Raw)
+	if issues != nil {
+		go func() {
+			for _, issue := range issues {
+				key, num := strings.ToUpper(issue[0]), issue[1]
+				_, found := projects[key]
+				if found {
+					result.Message <- url + key + "-" + num
+				}
+			}
+			result.Done <- true
+		}()
+	} else {
+		result.Done <- true
 	}
 
-	return "", nil
+	return result, nil
 }
 
 func init() {
+	var err error
 	url = os.Getenv(env)
-	bot.RegisterPassiveCommand(
+	jiraUser = os.Getenv(userEnv)
+	jiraPass = os.Getenv(passEnv)
+	baseURL = os.Getenv(baseURLEnv)
+
+	tp := gojira.BasicAuthTransport{
+		Username: jiraUser,
+		Password: jiraPass,
+	}
+
+	client, err = gojira.NewClient(tp.Client(), baseURL)
+	if err != nil {
+		panic(err)
```

Comment:
@sochotnicky it is not a good idea, this probrably would prevent the bot from booting up. I think it's a better ideia to just log the error and do not register the command.

Reply:
I can tweak though I actually did this on purpose, thinking that if this fails the configuration is seriously borked and it's better to crash during startup rather than fail during operation (i.e. fail fast). 

I will change it